duckduckgo/Android
# Back navigation to previous tab

Task/Issue URL: https://app.asana.com/0/414730916066338/1187653972293689/f
Tech Design URL: 
CC: 

**Description**:
Tries to fix the following issue:

Current behavior:
- User clicks on a link from tab A
- New tab is opened (Tab B)
- User presses back
- Home page shown (in Tab B)
- User presses back again (in Tab B)
- App is closed

Expected behavior:
- User clicks on a link from tab A
- New tab is opened (Tab B)
- User presses back (Tab B)
- Tab B closed
- Tab A opened

**Related open issues:**
https://github.com/duckduckgo/Android/issues/815
https://github.com/duckduckgo/Android/issues/696

**Steps to test this PR**:

Test: Open link on a new tab
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. open on a new tab (Tab B)
1. ensure new tab is opened with expected link
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link on a new tab (source tab deleted)
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. open on a new tab (Tab B)
1. ensure new tab is opened with expected link
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

Test: Open link on a background tab
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. Click on "Open in Background Tab" (Tab B)
1. ensure new tab is created (tabs counter)
1. Go to that tab (open tabswitcher, and change tab)
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link on a background tab (source tab deleted)
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. Click on "Open in Background Tab" (Tab B)
1. ensure new tab is created (tabs counter)
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

Test: Open link in new tab (target=_blank)
1. Open the app
1. Visit https://www.w3schools.com/code/tryit.asp?filename=GGETHXPDSEMN (Tab A)
1. Press run and click on "test" or "link in new tab"
1. ensure new tab is opened (Tab B)
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link in new tab (target=_blank / source tab deleted)
1. Open the app
1. Visit https://www.w3schools.com/code/tryit.asp?filename=GGETHXPDSEMN (Tab A)
1. Press run and click on "test" or "link in new tab"
1. ensure new tab is opened (Tab B)
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

**Steps to test previous logic is preserved**:

Test: Open a New Empty tab
1. Open the app
1. create a new tab manually from overflow menu or tabswitcher
1. Ensure new tab is created
1. perform a search o visit a website
1. press back
1. Ensure home screen is shown

Test: External links
1. Go to another app where you have a link to click
1. Open that link with DDG
1. Ensure link is opened in a new tab
1. Press back
1. Ensure previous app is opened again
1. Open DDG app again via task manager
1. Press back on the tab you previously opened
1. Ensure App gets closed

Test: Search widget
1. Click on our search widget
1. search something or click any suggestion
1. Ensure new tab is created with your search term
1. Press back
1. Ensure App gets closed

Test: Shortcut
1. Create a shortcut to any website
1. Click on the shortcut
1. Ensure new tab is created with your search term
1. Press back
1. Ensure App gets closed

**Test database migration**:
1. Install current `main` branch
1. Open as many tabs as you want
1. Install this branch
1. Open app again
1. Ensure app doesn't crash

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
